Tasks/AzureIoTEdge/util.ts

```diff
@@ -0,0 +1,238 @@
+    if (tl.getVariable(Constants.iotedgedevLockVersionKey)) {
+      version = tl.getVariable(Constants.iotedgedevLockVersionKey);
+    }
+    tl.debug(`The specified iotedgedev version is: ${version}`);
+    if (tl.osType() === Constants.osTypeLinux) {
+      cmds = [
+        { path: `sudo`, arg: `apt-get update`, execOption: Constants.execSyncSilentOption },
+        { path: `sudo`, arg: `apt-get install -y python-setuptools`, execOption: Constants.execSyncSilentOption },
+        { path: `sudo`, arg: `pip install ${Constants.iotedgedev}==${version}`, execOption: Constants.execSyncSilentOption },
+      ]
+    } else if (tl.osType() === Constants.osTypeWindows) {
+      cmds = [
+        { path: `pip`, arg: `install ${Constants.iotedgedev}==${version}`, execOption: Constants.execSyncSilentOption },
+      ]
+    }
+
+    try {
+      for (let cmd of cmds) {
+        let result = tl.execSync(cmd.path, cmd.arg, cmd.execOption);
+        if (result.code !== 0) {
+          tl.debug(result.stderr);
+        }
+      }
+    } catch (e) {
+      // If exception, record error message to debug
+      tl.debug(e);
+    }
+
+    let result = tl.execSync(`${Constants.iotedgedev}`, `--version`, Constants.execSyncSilentOption);
+    if (result.code === 0) {
+      console.log(tl.loc('DependencyInstallSuccess', Constants.iotedgedev, result.stdout.substring(result.stdout.indexOf("version"))));
+    } else {
+      throw Error(tl.loc('DependencyInstallFail', Constants.iotedgedev));
+    }
+  }
+
+  public static debugOsType() {
+    let cmd: string[] = null;
+    if (tl.osType() === Constants.osTypeWindows) {
+      cmd = ['systeminfo', null];
+    } else if (tl.osType() === Constants.osTypeLinux) {
+      cmd = [`lsb_release`, `-a`];
+    }
+    if (cmd != null) {
+      try {
+        let result = tl.execSync(cmd[0], cmd[1], Constants.execSyncSilentOption);
+        tl.debug(`OS is ${result.stdout}`);
+      } catch (e) {
+        tl.debug(`Error happened when fetching os info: ${e.message}`);
+      }
+    }
+  }
+
+  // test
+  // a b false
+  // docker.io docker.io true
+  // "docker.io","http://index.docker.io/v1" true
+  // "zhiqing.azurecr.io","http://zhiqing.azurecr.io" true
+  // "zhiqing.azurecr.io","https://zhiqing.azurecr.io" true
+  // "zhiqing.azurecr.io","https://zhiqing.azurecr.io/" true
+  public static isDockerServerMatch(a: string, b: string): boolean {
+    if (a === b) return true;
+    if (a.includes(Constants.defaultDockerHubHostname) && b.includes(Constants.defaultDockerHubHostname)) return true;
+
+    let reg = new RegExp(/^(?:https?:\/\/)?(.*?)\/?$/);
+    let aMatch = reg.exec(a);
+    let bMatch = reg.exec(b);
+    if (aMatch == null || bMatch == null) return false;
+    return aMatch[1] === bMatch[1];
+  }
+
+  // Check if self(task) is included in a build pipeline
+  public static checkSelfInBuildPipeline(): boolean {
+    let hostType = tl.getVariable('system.hostType').toLowerCase();
```

Comment:
are we differentiating task behaviour based on whether the task is included in CI or CD?

Reply:
Yes. But we didn't use different code logic to treat it. We just print out this information for help user debug issues.